src/Model bias analysis.ipynb

```diff
@@ -0,0 +1,1554 @@
+    "\n",
+    "texts = ['i am a {} person'.format(term) for term in TEST_TERMS]\n",
+    "scored_texts = zip(model.score(texts), texts)\n",
+    "for (score, text) in sorted(scored_texts):\n",
+    "    print '{:.3f}  {}'.format(score, text)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Bias-fuzzed test sets\n",
+    "\n",
+    "We compare AUC performance on a sample from the original test set. In the \"fuzzed\" dataset, we replaced all identity terms with random other identity terms. The \"non-fuzzed\" dataset is just the original examples.\n",
+    "\n",
+    "See [docs](https://github.com/conversationai/unintended-ml-bias-analysis#fuzzed-test-set) for more details on this dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 82,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def load_fuzzed_and_orig(model):\n",
+    "    fuzzed, nonfuzzed = (pd.read_csv('../eval_datasets/toxicity_fuzzed_testset.csv'),\n",
+    "                         pd.read_csv('../eval_datasets/toxicity_nonfuzzed_testset.csv'))\n",
+    "    fuzzed['score'] = model.score(fuzzed['comment'])\n",
+    "    nonfuzzed['score'] = model.score(nonfuzzed['comment'])\n",
+    "    return fuzzed, nonfuzzed"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 84,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 48.1 s, sys: 24.3 s, total: 1min 12s\n",
+      "Wall time: 7.96 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "fuzzed, nonfuzzed = load_fuzzed_and_orig(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### AUC\n",
+    "\n",
+    "We see that AUC on the fuzzed dataset is **0.944**, less than **0.957** on the original dataset. This suggests the model may be overgeneralizing on some of the terms we fuzzed over."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 88,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "0.95749598070739539"
+      ]
+     },
+     "execution_count": 88,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "compute_auc(nonfuzzed.toxic, nonfuzzed.score)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 87,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "0.94402486775230798"
+      ]
+     },
+     "execution_count": 87,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "compute_auc(fuzzed.toxic, fuzzed.score)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Largest score differences\n",
+    "\n",
+    "Qualitatively, it seems that the largest score differences hinge on the words \"gay\" and \"queer\"."
```

Comment:
We could make this quantitative by splitting the comments by which TEST_TERM they contain and then looking at the average delta.

Reply:
Yeah, we also want to compute per-term AUCs for the bias madlibs. I've made that change above.

I'll add a TODO for this one, since it's not as clear if we want to include this bias-fuzzing technique in the paper.

Reply:
SGTM